New Processing Algorithm to generate raster XYZ tiles

marcel-dancak opened the pull request:
## Description
This is a new processing tool for exporting map as XYZ raster tiles. It supports directory structure and MBTiles output format. Tiles can be stored in PNG or JPG image format.


Sponsored by:

- Hansestadt Herford (www.Herford.de)
- "SWK mbH" (https://swk.herford.de/)
- d.b.g. Datenbankgesellschaft mbH (https://datenbankgesellschaft.de/)

Commit:
Initial version of xyz tiles export (processing tool)

m-kuhn commented:
How about merging this with the rasterize map algorithm?

Commit:
Registration of TilesXYZ algorithm

wonder-sk commented:
> How about merging this with the rasterize map algorithm?

I don't think there is enough overlap... Both algorithms render a map, but the main focus here is to 1) output XYZ tiles (web mercator), 2) output multiple zoom levels, 3) support specific formats (directory / MBTiles), 4) use metatiling. And the output of this algorithm is not a single raster layer - it can be a directory tree etc.

m-kuhn commented:
Ok then.
A combination of any CRS with multiple zoom levels and metatiles - optionally in any raster file supporting pyramids - would have sounded like a nice use case to me.

Commit:
Commented out code for debugging (saving of metatiles), fixed python formatting

saberraz commented:
Some good points by @luipir:

> IMHO the game changes is to allow:
> 1) metatile rendering (as in your code and commonly done)
> 2) async tile writing 
> 3) parallel renderer instead of sync renderer gaining 50% in rendering time for complex stuffs.

Not sure if we are able to cover them within this work. But good to have it documented here for future reference, when we revisit the additional features.

luipir commented:
> Some good points by @luipir:

I already did this kind of work (but not integrated in processing) for a public institution... but unfortunately they did not published the code :( so I can't share. but they publish the city cadastre data directly from a qgis project. Rendering takes hours so it's important any further optimization.
QTile3 already have parallel writing (and seems that writer of this PR comes from that plugin). What I wasn't able to fix is a rendering problem when using ParallelRenderingJob that does not generate some styles, but it could be a problem introduced by my standalone application (do not have resources to investigate more).

Gustry commented:
Nice addition! 
The help, in the YML file, is missing from the PR for now.

nyalldawson commented on python/plugins/processing/algs/qgis/TilesXYZ.py:
```diff
@@ -0,0 +1,366 @@
+        return 'rastertools'
+
+    def initAlgorithm(self, config=None):
+        self.addParameter(QgsProcessingParameterString(self.NAME, self.tr('Name'), defaultValue='Test'))
```
Change "test" to something else

nyalldawson commented on python/plugins/processing/algs/qgis/TilesXYZ.py:
```diff
@@ -0,0 +1,366 @@
+        self.addParameter(QgsProcessingParameterFolderDestination(self.OUTPUT_DIRECTORY, self.tr('Output directory')))
+
+        self.addParameter(QgsProcessingParameterExtent(self.EXTENT, self.tr('Extent')))
+        self.addParameter(QgsProcessingParameterNumber(self.ZOOM_MIN,
```
These params need a min/max value range set

nyalldawson commented on python/plugins/processing/algs/qgis/TilesXYZ.py:
```diff
@@ -0,0 +1,366 @@
+        self.addParameter(QgsProcessingParameterNumber(self.ZOOM_MAX,
+                                                       self.tr('Maximum zoom'),
+                                                       defaultValue=12))
+        self.addParameter(QgsProcessingParameterNumber(self.DPI, self.tr('DPI'), defaultValue=96))
```
Same here -- a dpi <1 should not be permitted

marcel-dancak replied:
All numeric parameters should now have a reasonable min/max values set.

nyalldawson commented on python/plugins/processing/algs/qgis/TilesXYZ.py:
```diff
@@ -0,0 +1,366 @@
+        tile_width = 256
+        tile_height = 256
+
+        project = QgsProject.instance()
```
Take the project from context, don't use QgsProject.instance()

nyalldawson commented on python/plugins/processing/algs/qgis/TilesXYZ.py:
```diff
@@ -0,0 +1,366 @@
+        tile_height = 256
+
+        project = QgsProject.instance()
+        visible_layers = [item.layer() for item in project.layerTreeRoot().findLayers() if item.isVisible()]
```
This isn't safe to run in a background thread -- it needs to be moved to a prepareAlgorithm step and run in the main thread.

marcel-dancak replied:
I have moved accessing of project's layers into the prepareAlgorithm. If some other parts will need to be moved as well, I think @wonder-sk will take care of it.

nyalldawson commented on python/plugins/processing/algs/qgis/TilesXYZ.py:
```diff
@@ -0,0 +1,366 @@
+                image.setDotsPerMeterX(dpm)
+                image.setDotsPerMeterY(dpm)
+                painter = QPainter(image)
+                job = QgsMapRendererCustomPainterJob(settings, painter)
```
I don't think this is safe to do from a background thread -- I think you need to setup ALL the jobs in advance, in the prepareAlgorithm step

luipir replied:
@nyalldawson this would mean allocating all tiles at different zoom level in advance! for a mobile project ti may work, but not rendering a big area. I've experience with Barcelona municipality and it generate many GBs from 10 to 21 zoom level. Any clue to prepare jobs without this memory allocatio? something like a job queue prepared in the main thread and the processor/renderer as consumer?

wonder-sk replied:
Yeah I was thinking about the same - right now we don't have a way to modify map settings of an existing map renderer job and creating great amount of these in advance is probably going to be very wasteful of resources. Looking at the Rasterize algorithm, it also creates QgsMapRendererCustomPainterJob for each tile it renders within processAlgorithm.

While I fully agree this is suboptimal and potentially unsafe, it works fine and it's probably better to keep it this way until there's API to handle this...

nyalldawson commented:
Looking good, nice feature!

My strong preference would be that any new algorithm, especially one as complex as this, is written in c++ and not Python. The Python algorithms have just proved too fragile in the past, and we very much rely on compile time checks to ensure that they don't break. Just something for (everyone) to keep in mind for future work -- in the meantime, you'll need to soak this one with tests :stuck_out_tongue_winking_eye:

Commit:
MBTilesWriter can write tiles in any order, proper transforming of extents

wonder-sk commented:
> My strong preference would be that any new algorithm, especially one as complex as this, is written in c++ and not Python. The Python algorithms have just proved too fragile in the past, and we very much rely on compile time checks to ensure that they don't break.

Agreed. Here the main reason to do it in Python was the intention to provide this algorithm as a processing provider in a plugin for QGIS 3.4 and 3.6. I would be happy to move it to c++ post 3.8 release together with safety improvements you were concerned about.

Commit:
Adjusted min/max/default values of parameters, moved some code into 'prepareAlgorithm'

Commit:
Added help text

wonder-sk committed:
[processing] Add 'project' to test definition and 'directory' output test

- for algorithms that produce directory output, it is possible to test
  that directory contents are exactly the same (recursively)
- added possibility to have a project file loaded before an algorithm is run
- documented the new additions (+ few existing ones)

wonder-sk committed:
Remove NAME parameter, use OUTPUT_FILE instead

For output to directory, OUTPUT_DIRECTORY destination parameter is used.
For output to MBTiles file, OUTPUT_FILE destimation parameter is used.

wonder-sk committed:
Add test for XYZ Tiles algorithm

wonder-sk committed:
Raise an exception if the necessary parameter was not specified

Gustry commented on python/plugins/processing/algs/help/qgis.yaml:
```diff
@@ -509,6 +509,13 @@ qgis:sumlinelengths: >
+
+  Tile images can be saved as individual images in directory structure, or as single file in MBTiles format.
+
+  Tile size is fixed to 256x256.
```
Some information about the output projection would be nice IMHO. Does it always output 3857? I guess yes, but it's good to mention.

BTW, the algorithm is missing tags. I was looking 'mbtiles' in the processing toolbox without finding it for example.

Nice addition, thanks.

dhelie commented:
that's great. Upon testing i noticed that it is producing pngs with white backgrounds.
adding settings.setBackgroundColor(QColor(Qt.transparent)) to QgsMapSettings() fixes the issue

isghj5 commented:
I missed the part where this is getting converted to C++ soon, but I got a basic multithreading working with the python version:

In TilesXYZ.py, I took most of the looping code and moved it to a new function, then mapped the function to all the tiles through ThreadPoolExecutor, which handled all the threading for me, except the settings, to avoid threads fighting over settings I made settings a dictionary, one settings per thread.

In my loose test, with the same extent, zoom, image quality, I dropped from 53 seconds to 15 seconds with my quad core, I'm sure people rendering tiles on new beefier ryzen/threadripper/xeon machines will see a bigger boost from this.
![speed boost](https://user-images.githubusercontent.com/42048411/60306622-c02b8d80-98f5-11e9-92b2-f7b18a5233bc.png)

I'm bad with UI design, not sure how to add a core count option to the UI so the user can change it if they want, this code detects the max from the computer and uses that, but you can just replace cpu_count() with a number if you want in the code.

Known bug: if you hit close before hitting cancel while it's rendering, the threads will stay alive in the background spinning. Not sure where the cancel button code is.

[TilesXYZ.diff.txt](https://github.com/qgis/QGIS/files/3336938/TilesXYZ.diff.txt)

[TilesXYZ.py.txt](https://github.com/qgis/QGIS/files/3336939/TilesXYZ.py.txt)

PeterPetrik commented:
@isghj5 Would it be possible to create separate pull request with your addition and discuss there? looks like valuable addition

DelazJ commented:
@Gustry 
> The help, in the YML file, is missing from the PR for now.

And it has been merged without (and without labels to generate the issue report in docs repo btw). I think a short help in the dialog for each new algorithm and expression function should be a requirement. This helps users and docs writers.

wonder-sk commented:
@Gustry @DelazJ apologies for ignoring the note about update of YML file and thanks for the reminder - will try to fix that

Gustry commented:
A description in the YML has been added, no? I can see it's the first file in the PR.
But I think it was tags in the alg which were missing.

DelazJ commented:
@wonder-sk  We were documenting it in the docs so had to do some more research but no worries. The issue is that for some reason, they do not display in the GUI. Thanks
